anjali/sb-pt-cont/9.js

```diff
@@ -0,0 +1,231 @@
+      uniqueStartingSpriteCostumes.push(costume);
+    }
+  }
+  if (spriteIds.length == uniqueStartingSpriteCostumes.length) {
+    validationProps.successCriteria.differentCostumes = true;
+  }
+}
+
+// Check if student is interacting with the program
+// version 1
+for (var spriteId in spriteIds) {
+  if (getProp({id: spriteId}, "timeout")>0) {
+    validationProps.waitTime = World.frameCount + 60;
+    break;
+  } else {
+    validationProps.waitTime = validationProps.previous.waitTime;
+  }
+}
+
+/*
+// version 2
+if (validationProps.successTime && (eventLog.length > validationProps.previous.eventLogLength)) {
+  validationProps.waitTime = World.frameCount + 100;
+} else {
+  validationProps.waitTime = validationProps.previous.waitTime;
+}
+*/
+
+// Keeps track of which sprites were clicked on
+if(!validationProps.spritesClickedOn) {
+  validationProps.spritesClickedOn = [];
+}
+
+// Keeps track of which sprites caused speech on click
+if(!validationProps.spriteClicksCauseSpeech) {
+  validationProps.spriteClicksCauseSpeech = [];
+}
+
+// Helper function
+// returns true if spriteId is in spriteIdArray
+function member(spriteId, spriteIdArray) {
+  for (var i = 0; i < spriteIdArray.length; i++) {
+    if (spriteIdArray[i] == spriteId) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// New event happened
+if (eventLog.length > validationProps.previous.eventLogLength) {
+  // check eventLog for WhenClick or WhileClick event
+  var currentEvent = eventLog[eventLog.length - 1];
+
+  if (currentEvent.includes("whenClick: ") || currentEvent.includes("whileClick: ")) {
+    // current event is a click
+    var clickedSpriteId = parseInt(currentEvent.split(" ")[1]);
+    validationProps.spritesClickedOn.push(clickedSpriteId);
+    validationProps.successCriteria.numSpritesClicked = validationProps.spritesClickedOn.length;
+    if (!member(clickedSpriteId, validationProps.spriteClicksCauseSpeech)) {
+      var newSpriteSpeaking = false;
+      for (var spriteId in spriteIds) {
+        if (getProp({id: spriteId}, "speech") && getProp({id: spriteId}, "timeout")==120) {
+          // new sprite caused speech in some sprite
+          newSpriteSpeaking = true;
+          break;
+        }
+      }
+      if (newSpriteSpeaking) {
+        validationProps.spriteClicksCauseSpeech.push(clickedSpriteId);
+      	validationProps.successCriteria.numSpriteClicksCauseSpeech = validationProps.spriteClicksCauseSpeech.length;
+      }
+    }
+  }
+}
+
+// Store previous values
+validationProps.previous.eventLogLength = eventLog.length;
+validationProps.previous.waitTime = validationProps.waitTime;
+
+// Set success time if success
+if (validationProps.successCriteria.threeSprites &&
+    validationProps.successCriteria.differentCostumes &&
+    validationProps.successCriteria.differentLocations &&
+    validationProps.successCriteria.noSpritesTouching &&
+    validationProps.successCriteria.numSpritesClicked>=validationProps.previous.numStartingSprites &&
```

Comment:
This and the next line should be `>= 3` I think?

Reply:
Make the validation just check for 3. don't change instructions. ask mike for future.